src/python/pants/engine/target.py

```diff
@@ -1624,20 +1625,34 @@ def any_are_covered_by_includes(self, addresses: Tuple[Address, ...]) -> bool:
         )
 
     @memoized_method
-    def remaining_after_ignores(self, addresses: Tuple[Address, ...]) -> frozenset[Address]:
-        """All addresses that are not covered by the explicitly provided ignores.
+    def remaining_after_disambiguation(
+        self, addresses: Tuple[Address, ...], owners_must_be_ancestors: bool
+    ) -> frozenset[Address]:
+        """All addresses that remain after ineligible candidates are discarded.
 
-        Note that if the input addresses are file addresses, they will still be marked as covered if
-        their original BUILD target is in the explicitly provided ignores.
+        Candidates are removed if they appear as ignores (`!` and `!!)` in the `dependencies`
+        field. Note that if the input addresses are file addresses, they will still be marked as
+        covered if their original BUILD target is in the explicitly provided ignores.
+
+        Candidates are also removed if `owners_must_be_ancestors` is True and the targets are not
+        ancestors, e.g. `root2:tgt` is not a valid candidate for something defined in `root1`.
```

Comment:
It seems like what we're doing here is converting into a module (which discards the parent directory in `root1/app.py`, `root2/app.py`), and then adding a heuristic to ensure that we use the "right" module. It seems like an alternative would be to ensure that before inferring, we absolutize the file path that was specified, and extract a dep on the entire file path rather than just on the module.

Is this because dep inference always extracts "modules" rather than files? This feels like it will be relevant to @tdyas's inference for Terraform, which will also extract relative file deps.

Reply:
Python is specifically converting into a module, but that is not true for other implementations like Shell and Protobuf. Python uses module because almost always, that is the right abstraction to deal with, given how Python imports work.

Reply:
And it's not worth it allow both literal-files and modules to be inferred within the python inference rules, to avoid the need for special case code in the general inference rules? Or do you think that this `owners_must_be_ancestors` flag is going to be useful in other cases?

Reply:
> And it's not worth it allow both literal-files and modules to be inferred within the python inference rules, to avoid the need for special case code in the general inference rules?

No, I don't think so. The Python code is already really complex.

> Or do you think that this owners_must_be_ancestors flag is going to be useful in other cases?

I'm not certain. But it's defaulted to False and also well tested - most call sites can ignore this.